### ISSUE TITLE
Is FA3 less accurate than FA2 in bf16 computation?

### ISSUE THREAD

**Issue by complexfilter** [Dao-AILab/flash-attention#1138]
I did some numerical test for FA3 vs FA2, using PyTorch's `scaled_dot_product_attention` result as the reference. I use the relative error in Frobenius norm as the metric. I found that when the reference results were computed in float64 or float32 precision, FA3 and FA2 had the same relative error to the reference. 

However, when the reference results were computed in bfloat16, FA2's relative error is 1/20 of FA3's relative error. Is this an inherent drawback of FA3 or this is just a fixable bug?

Source code to reproduce:

```
import torch
import torch.nn.functional as F
from flash_attn.flash_attn_interface import flash_attn_varlen_func, flash_attn_func
import flashattn_hopper_cuda

def get_tensors(batch_size, seq_len, head_size, dim):
    torch.manual_seed(42)
    q = torch.randn((batch_size, seq_len, head_size, dim), dtype=torch.bfloat16, device="cuda", requires_grad=True)
    k = torch.randn((batch_size, seq_len, head_size, dim), dtype=torch.bfloat16, device="cuda", requires_grad=True)
    v = torch.randn((batch_size, seq_len, head_size, dim), dtype=torch.bfloat16, device="cuda", requires_grad=True)
    return q, k, v

class FA3_forward_FA3_backward(torch.autograd.Function):
    @staticmethod
    def forward(ctx, q, k, v):
        ctx.softmax_scale = q.shape[-1] ** (-0.5)   
        out, _, _, _, _, lse, _ = flashattn_hopper_cuda.fwd(q, k, v, None, ctx.softmax_scale, False)
        ctx.save_for_backward(out, lse, q, k, v)
        return out

    @staticmethod
    def backward(ctx, grad_output):
        # Use the saved input from the forward pass
        out, lse, q, k, v = ctx.saved_tensors
        dq, dk, dv = torch.empty_like(q), torch.empty_like(k), torch.empty_like(v)
        dq, dk, dv, *rest = flashattn_hopper_cuda.bwd(grad_output, q, k, v, out, lse, dq, dk, dv, ctx.softmax_scale, False, False)
        return dq, dk, dv

if __name__=="__main__":
    batch_size, seq_len, head_size, dim = 4, 8192, 8, 128
    print(f"batch_size: {batch_size},  seq_len: {seq_len}, head_size: {head_size}, dim: {dim}")
    q, k, v = get_tensors(batch_size, seq_len, head_size, dim)
    q_ref = q.transpose(1, 2)
    k_ref = k.transpose(1, 2)
    v_ref = v.transpose(1, 2)
    out_ref = F.scaled_dot_product_attention(q_ref, k_ref, v_ref)
    out_ref = out_ref.transpose(1, 2)

    q_fa3 = q.type(torch.bfloat16)
    k_fa3 = k.type(torch.bfloat16)
    v_fa3 = v.type(torch.bfloat16)
    out_fa3 = FA3_forward_FA3_backward.apply(q_fa3, k_fa3, v_fa3)

    q_fa2 = q_fa3.detach().clone().requires_grad_(True)
    k_fa2 = k_fa3.detach().clone().requires_grad_(True)
    v_fa2 = v_fa3.detach().clone().requires_grad_(True)        
    out_fa2 = flash_attn_func(q_fa2, k_fa2, v_fa2)
    print(torch.norm(out_fa2.type(out_ref.dtype) - out_ref, p='fro') / torch.norm(out_ref, p='fro'))
    print(torch.norm(out_fa3.type(out_ref.dtype) - out_ref, p='fro') / torch.norm(out_ref, p='fro'))
```

**Comment by tridao** [comment]
:D becuase `scaled_dot_product_attention` is essnetially runnign FA2 code.

**Comment by tridao** [comment]
what you're observing is that output FA2 code with a slightly different compiler flags (e..g `scaled_dot_product_attention`) is closer to FA2 output compared to FA3 output :D

**Comment by complexfilter** [comment]
Thanks for the explanation! I wrote a naive scaled dot product attention function, I cannot find significant error difference across FA2 and FA3.

```
import torch
from flash_attn.flash_attn_interface import flash_attn_varlen_func, flash_attn_func
import flashattn_hopper_cuda
import math

def get_tensors(batch_size, seq_len, head_size, dim):
    q = torch.randn((batch_size, seq_len, head_size, dim), dtype=torch.bfloat16, device="cuda", requires_grad=True)
    k = torch.randn((batch_size, seq_len, head_size, dim), dtype=torch.bfloat16, device="cuda", requires_grad=True)
    v = torch.randn((batch_size, seq_len, head_size, dim), dtype=torch.bfloat16, device="cuda", requires_grad=True)
    return q, k, v


def scaled_dot_product_attention(query, key, value):
    scale_factor = 1 / math.sqrt(query.size(-1)) 
    attn_weight = query @ key.transpose(-2, -1) * scale_factor
    attn_weight = torch.softmax(attn_weight, dim=-1)
    return attn_weight @ value
    

class FA3_forward_FA3_backward(torch.autograd.Function):
    @staticmethod
    def forward(ctx, q, k, v):
        ctx.softmax_scale = q.shape[-1] ** (-0.5)   
        out, _, _, _, _, lse, _ = flashattn_hopper_cuda.fwd(q, k, v, None, ctx.softmax_scale, False)
        ctx.save_for_backward(out, lse, q, k, v)
        return out

    @staticmethod
    def backward(ctx, grad_output):
        # Use the saved input from the forward pass
        out, lse, q, k, v = ctx.saved_tensors
        dq, dk, dv = torch.empty_like(q), torch.empty_like(k), torch.empty_like(v)
        dq, dk, dv, *rest = flashattn_hopper_cuda.bwd(grad_output, q, k, v, out, lse, dq, dk, dv, ctx.softmax_scale, False, False)
        return dq, dk, dv

if __name__=="__main__":
    batch_size, seq_len, head_size, dim = 4, 8192, 8, 128
    print(f"batch_size: {batch_size},  seq_len: {seq_len}, head_size: {head_size}, dim: {dim}")
    q, k, v = get_tensors(batch_size, seq_len, head_size, dim)
    q_ref = q.transpose(1, 2)
    k_ref = k.transpose(1, 2)
    v_ref = v.transpose(1, 2)
    out_ref = scaled_dot_product_attention(q_ref, k_ref, v_ref)
    out_ref = out_ref.transpose(1, 2)

    q_fa3 = q.type(torch.bfloat16)
    k_fa3 = k.type(torch.bfloat16)
    v_fa3 = v.type(torch.bfloat16)
    out_fa3 = FA3_forward_FA3_backward.apply(q_fa3, k_fa3, v_fa3)

    q_fa2 = q_fa3.detach().clone().requires_grad_(True)
    k_fa2 = k_fa3.detach().clone().requires_grad_(True)
    v_fa2 = v_fa3.detach().clone().requires_grad_(True)        
    out_fa2 = flash_attn_func(q_fa2, k_fa2, v_fa2)
    print(torch.norm(out_fa2.type(out_ref.dtype) - out_ref, p='fro') / torch.norm(out_ref, p='fro'))
    print(torch.norm(out_fa3.type(out_ref.dtype) - out_ref, p='fro') / torch.norm(out_ref, p='fro'))
```

**Comment by nighting0le01** [comment]
can we swap the attention layers of a model trained with FA2 (SDPA) directly with FA3? without training? @complexfilter
